File: frontend/src/types/interactive.ts
/**
 * TypeScript type definitions for interactive podcast conversations.
 *
 * These types support the chat panel, voice input/output, and
 * WebSocket streaming for real-time conversations.
 */

// =============================================================================
// Enums
// =============================================================================

/**
 * Role of a chat message participant.
 */
export type MessageRole = 'user' | 'assistant' | 'system'

/**
 * Types of WebSocket streaming events for interactive chat.
 */
export type StreamEventType =
  | 'assistant_start'
  | 'assistant_chunk'
  | 'assistant_end'
  | 'audio_ready'
  | 'error'
  | 'session_started'
  | 'session_ended'
  | 'heartbeat'
  | 'pong'

/**
 * Types of messages sent from client to server.
 */
export type ClientMessageType = 'message' | 'ping' | 'end_session'

// =============================================================================
// Chat Message Types
// =============================================================================

/**
 * A single message in the conversation.
 */
export interface ChatMessage {
  /** Unique message identifier */
  id: string
  /** Who sent the message */
  role: MessageRole
  /** Text content of the message */
  content: string
  /** When the message was sent */
  timestamp: string
  /** TTS audio URL for playback (assistant messages only) */
  audio_url?: string
}

// =============================================================================
// Session Types
// =============================================================================

/**
 * Response after starting a session.
 */
export interface SessionResponse {
  /** Unique session identifier */
  session_id: string
  /** Associated podcast job ID */
  job_id: string
  /** Session creation time */
  created_at: string
  /** Initial greeting from assistant */
  welcome_message: ChatMessage
}

/**
 * Information about an active session.
 */
export interface SessionInfo {
  /** Session identifier */
  session_id: string
  /** Associated podcast job ID */
  job_id: string
  /** When session started */
  created_at: string
  /** Number of messages exchanged */
  message_count: number
  /** Whether session is still active */
  is_active: boolean
}

// =============================================================================
// Message Request/Response Types
// =============================================================================

/**
 * Request to send a text message.
 */
export interface MessageRequest {
  /** User's message text */
  content: string
  /** Whether to generate TTS response audio */
  generate_audio?: boolean
}

/**
 * Response to a user message (non-streaming).
 */
export interface MessageResponse {
  /** The user's message that was sent */
  user_message: ChatMessage
  /** The AI's response */
  assistant_message: ChatMessage
}

/**
 * Response from voice transcription.
 */
export interface TranscriptionResponse {
  /** Transcribed text from audio */
  transcription: string
  /** Transcription confidence score (0-1) */
  confidence: number
  /** Audio duration in seconds */
  duration_seconds?: number
}

// =============================================================================
// History Types
// =============================================================================

/**
 * Conversation history for a session.
 */
export interface HistoryResponse {
  /** Session identifier */
  session_id: string
  /** List of messages in chronological order */
  messages: ChatMessage[]
  /** Total message count */
  total: number
}

// =============================================================================
// WebSocket Stream Types
// =============================================================================

/**
 * WebSocket message for real-time streaming.
 */
export interface StreamMessage {
  /** Event type */
  type: StreamEventType
  /** Session this event belongs to */
  session_id: string
  /** Message ID (for chunk events) */
  message_id?: string
  /** Text content (for chunk events) */
  content?: string
  /** Audio URL (for audio_ready events) */
  audio_url?: string
  /** Error message (for error events) */
  error?: string
  /** Timestamp */
  timestamp?: string
}

/**
 * Message from WebSocket client to server.
 */
export interface ClientMessage {
  /** Message type */
  type: ClientMessageType
  /** Message content (for message type) */
  content?: string
  /** Whether to generate TTS */
  generate_audio?: boolean
}

// =============================================================================
// Voice Recording Types
// =============================================================================

/**
 * Voice recording state.
 */
export interface VoiceRecordingState {
  /** Whether currently recording */
  isRecording: boolean
  /** Audio level (0-1) for visualizer */
  audioLevel: number
  /** Whether microphone permission is granted */
  hasPermission: boolean
  /** Recording duration in seconds */
  duration: number
}

/**
 * Voice recording configuration.
 */
export interface VoiceRecordingConfig {
  /** Audio MIME type */
  mimeType: string
  /** Sample rate in Hz */
  sampleRate: number
  /** Enable echo cancellation */
  echoCancellation: boolean
  /** Enable noise suppression */
  noiseSuppression: boolean
}

// =============================================================================
// Audio Playback Types
// =============================================================================

/**
 * Audio playback state.
 */
export interface AudioPlaybackState {
  /** Whether audio is currently playing */
  isPlaying: boolean
  /** Current playback position in seconds */
  currentTime: number
  /** Total duration in seconds */
  duration: number
  /** ID of currently playing message */
  messageId?: string
}

/**
 * Audio queue item for sequential playback.
 */
export interface AudioQueueItem {
  /** Message ID */
  messageId: string
  /** Audio URL */
  url: string
}

// =============================================================================
// Chat Panel Types
// =============================================================================

/**
 * Chat panel display mode.
 */
export type ChatPanelMode = 'collapsed' | 'expanded' | 'fullscreen'

/**
 * Chat input state.
 */
export interface ChatInputState {
  /** Current text in input */
  text: string
  /** Whether input is disabled */
  isDisabled: boolean
  /** Whether user is currently typing */
  isTyping: boolean
}

/**
 * Message being streamed (partial content).
 */
export interface StreamingMessage {
  /** Message ID */
  id: string
  /** Accumulated content so far */
  content: string
  /** Whether streaming is complete */
  isComplete: boolean
}

// =============================================================================
// Hook Return Types
// =============================================================================

/**
 * Return type for useInteractiveChat hook.
 */
export interface UseInteractiveChatReturn {
  /** Active session ID */
  sessionId: string | null
  /** All messages in conversation */
  messages: ChatMessage[]
  /** Whether WebSocket is connected */
  isConnected: boolean
  /** Whether waiting for response */
  isLoading: boolean
  /** Current streaming message (if any) */
  streamingMessage: StreamingMessage | null
  /** Error message (if any) */
  error: string | null
  /** Start a new session */
  startSession: (jobId: string) => Promise<void>
  /** Send a text message */
  sendMessage: (text: string) => Promise<void>
  /** End the current session */
  endSession: () => Promise<void>
  /** Clear error state */
  clearError: () => void
}

/**
 * Return type for useVoiceRecording hook.
 */
export interface UseVoiceRecordingReturn {
  /** Whether currently recording */
  isRecording: boolean
  /** Audio level (0-1) for visualizer */
  audioLevel: number
  /** Whether microphone permission is granted */
  hasPermission: boolean
  /** Recording duration in seconds */
  duration: number
  /** Start recording */
  startRecording: () => Promise<void>
  /** Stop recording and return audio blob */
  stopRecording: () => Promise<Blob | null>
  /** Request microphone permission */
  requestPermission: () => Promise<boolean>
  /** Error message (if any) */
  error: string | null
}

/**
 * Return type for useAudioPlayback hook.
 */
export interface UseAudioPlaybackReturn {
  /** Whether audio is currently playing */
  isPlaying: boolean
  /** ID of currently playing message */
  currentMessageId: string | null
  /** Play audio for a message */
  play: (url: string, messageId: string) => void
  /** Stop playback */
  stop: () => void
  /** Add to playback queue */
  queue: (url: string, messageId: string) => void
  /** Clear the queue */
  clearQueue: () => void
}

// =============================================================================
// API Response Types
// =============================================================================

/**
 * Generic API error response.
 */
export interface InteractiveErrorResponse {
  error: string
  message: string
  details?: Record<string, unknown>
}

File: backend/app/models/interactive.py
"""
Pydantic models for interactive podcast conversations.

These models define the structure for chat sessions, messages,
and WebSocket streaming events.
"""

from typing import Optional, List, Literal
from datetime import datetime
from enum import Enum
from pydantic import BaseModel, Field


class MessageRole(str, Enum):
    """
    Role of a chat message participant.

    Attributes:
        USER: Message from the user
        ASSISTANT: Response from the AI assistant
        SYSTEM: System-generated messages (e.g., session start)
    """
    USER = "user"
    ASSISTANT = "assistant"
    SYSTEM = "system"


class StreamEventType(str, Enum):
    """
    Types of WebSocket streaming events.

    Attributes:
        ASSISTANT_START: AI response starting
        ASSISTANT_CHUNK: Partial response text
        ASSISTANT_END: AI response complete
        AUDIO_READY: TTS audio available
        ERROR: An error occurred
        SESSION_STARTED: Session initialized
        SESSION_ENDED: Session terminated
    """
    ASSISTANT_START = "assistant_start"
    ASSISTANT_CHUNK = "assistant_chunk"
    ASSISTANT_END = "assistant_end"
    AUDIO_READY = "audio_ready"
    ERROR = "error"
    SESSION_STARTED = "session_started"
    SESSION_ENDED = "session_ended"


# ============================================================
# Chat Message Models
# ============================================================

class ChatMessage(BaseModel):
    """
    A single message in the conversation.

    Attributes:
        id: Unique message identifier
        role: Who sent the message
        content: Text content of the message
        timestamp: When the message was sent
        audio_url: URL to TTS audio (assistant messages only)
    """
    id: str = Field(..., description="Unique message identifier")
    role: MessageRole = Field(..., description="Message sender role")
    content: str = Field(..., description="Message text content")
    timestamp: datetime = Field(default_factory=datetime.utcnow)
    audio_url: Optional[str] = Field(None, description="TTS audio URL for playback")


# ============================================================
# Session Models
# ============================================================

class SessionRequest(BaseModel):
    """
    Request to start a new interactive session.

    Attributes:
        job_id: Job ID of the podcast to discuss
    """
    job_id: str = Field(..., description="Podcast job ID for context")


class SessionResponse(BaseModel):
    """
    Response after starting a session.

    Attributes:
        session_id: Unique session identifier
        job_id: Associated podcast job ID
        created_at: Session creation time
        welcome_message: Initial greeting from assistant
    """
    session_id: str = Field(..., description="Unique session identifier")
    job_id: str = Field(..., description="Associated podcast job ID")
    created_at: datetime = Field(default_factory=datetime.utcnow)
    welcome_message: ChatMessage = Field(..., description="Initial assistant greeting")


# ============================================================
# Message Request/Response Models
# ============================================================

class MessageRequest(BaseModel):
    """
    Request to send a text message.

    Attributes:
        content: User's message text
        generate_audio: Whether to generate TTS response audio
    """
    content: str = Field(..., min_length=1, max_length=2000, description="Message text")
    generate_audio: bool = Field(True, description="Generate TTS for response")


class MessageResponse(BaseModel):
    """
    Response to a user message (non-streaming).

    Attributes:
        user_message: The user's message that was sent
        assistant_message: The AI's response
    """
    user_message: ChatMessage = Field(..., description="User's sent message")
    assistant_message: ChatMessage = Field(..., description="AI response")


class VoiceMessageRequest(BaseModel):
    """
    Request metadata for voice message upload.

    Attributes:
        generate_audio: Whether to generate TTS response audio
    """
    generate_audio: bool = Field(True, description="Generate TTS for response")


class TranscriptionResponse(BaseModel):
    """
    Response from voice transcription.

    Attributes:
        transcription: Transcribed text from audio
        confidence: Transcription confidence score (0-1)
        duration_seconds: Audio duration
    """
    transcription: str = Field(..., description="Transcribed text")
    confidence: float = Field(..., ge=0, le=1, description="Confidence score")
    duration_seconds: Optional[float] = Field(None, description="Audio duration")


# ============================================================
# History Models
# ============================================================

class HistoryResponse(BaseModel):
    """
    Conversation history for a session.

    Attributes:
        session_id: Session identifier
        messages: List of messages in chronological order
        total: Total message count
    """
    session_id: str = Field(..., description="Session identifier")
    messages: List[ChatMessage] = Field(default_factory=list)
    total: int = Field(0, ge=0)


# ============================================================
# WebSocket Stream Models
# ============================================================

class StreamMessage(BaseModel):
    """
    WebSocket message for real-time streaming.

    Attributes:
        type: Event type
        session_id: Session this event belongs to
        message_id: Message ID (for chunk events)
        content: Text content (for chunk events)
        audio_url: Audio URL (for audio_ready events)
        error: Error message (for error events)
    """
    type: StreamEventType = Field(..., description="Event type")
    session_id: str = Field(..., description="Session identifier")
    message_id: Optional[str] = Field(None, description="Message ID")
    content: Optional[str] = Field(None, description="Text content")
    audio_url: Optional[str] = Field(None, description="Audio URL")
    error: Optional[str] = Field(None, description="Error message")
    timestamp: datetime = Field(default_factory=datetime.utcnow)


class ClientMessage(BaseModel):
    """
    Message from WebSocket client.

    Attributes:
        type: Message type (message, ping, end_session)
        content: Message content (for message type)
        generate_audio: Whether to generate TTS
    """
    type: Literal["message", "ping", "end_session"] = Field(..., description="Message type")
    content: Optional[str] = Field(None, description="Message content")
    generate_audio: bool = Field(True, description="Generate TTS for response")


# ============================================================
# Session Info Models
# ============================================================

class SessionInfo(BaseModel):
    """
    Information about an active session.

    Attributes:
        session_id: Session identifier
        job_id: Associated podcast job ID
        created_at: When session started
        message_count: Number of messages exchanged
        is_active: Whether session is still active
    """
    session_id: str
    job_id: str
    created_at: datetime
    message_count: int = 0
    is_active: bool = True
